modules/azure/src/main/clojure/xtdb/azure/object_store.clj

```diff
@@ -53,7 +54,44 @@
 (defn- delete-blob [^BlobContainerClient blob-container-client blob-name]
   (-> (.getBlobClient blob-container-client blob-name)
       (.deleteIfExists)))
-(defrecord AzureBlobObjectStore [^BlobContainerClient blob-container-client prefix ^NavigableSet file-name-cache ^AutoCloseable file-list-watcher]
+
+(def ^Base64$Encoder base-64-encoder (Base64/getEncoder))
+
+(defn block-number->base64-block-id [block-number]
+  (.encodeToString base-64-encoder (.getBytes (str block-number))))
+
+(defrecord MultipartUpload [^BlockBlobClient block-blob-client on-complete ^List !staged-block-ids]
+  IMultipartUpload
+  (uploadPart [_  buf]
+    (CompletableFuture/completedFuture
+     (let [block-number (inc (count !staged-block-ids))
+           block-id (block-number->base64-block-id block-number)
+           binary-data (BinaryData/fromByteBuffer buf)]
+       (.stageBlock block-blob-client block-id binary-data)
+       (.add !staged-block-ids block-id))))
+
+  (complete [_]
+    (CompletableFuture/completedFuture
+     (do
+       (.commitBlockList block-blob-client !staged-block-ids)
+       ;; Run passed in on-complete function (adds the key to the filename cache)
+       (on-complete))))
+  
+  (abort [_]
```

Comment:
Doesn't seem to be a direct way to delete uncomitted blocks other than this - using the same idea as from this StackOverflow answer: https://stackoverflow.com/questions/68662051/how-to-delete-uncommitted-blocks-in-azure-blob-storage

Reply:
(They will be automatically deleted after 7 days if not cleaned up manually, but I think we'd prefer to explicitly remove these failed blobs?)